### PR TITLE
[KernelGen][Nvidia] Add __irshift__ operator with Triton kernel

### PR DESCRIPTION
**Pull request by bwbwzzz**
## Summary
Adds a Triton kernel for `__irshift__` (inplace bitwise right shift, `a >>= b`). This operator performs element-wise right bit-shift on integer tensors, storing the result back into the input tensor. Triton kernel implementation for __irshift__.

## Testing
- Validated against PyTorch reference on device side
- Tested on: Nvidia, Tianshu, Muxi, Ascend, Hygon

## Performance
Test command: `pytest benchmark/test_irshift__.py --level core` (NVIDIA H20)

### torch.int16
| Status | Torch Latency (ms) | Gems Latency (ms) | Speedup | TFLOPS | Shape |
|---|---|---|---|---|---|
| SUCCESS | 1.522 | 1.463 | **1.040x** | 1.468 | [1073741824] |
| SUCCESS | 0.006 | 0.005 | **1.085x** | 0.002 | [64, 64] |
| SUCCESS | 0.030 | 0.030 | **1.006x** | 1.134 | [4096, 4096] |
| SUCCESS | 0.030 | 0.030 | **1.011x** | 1.130 | [64, 512, 512] |
| SUCCESS | 1.606 | 1.607 | **0.999x** | 1.336 | [1024, 1024, 1024] |
| **Int16 Mean** | — | — | **~1.028x** | — | 5 cases, all PASS |

### torch.int32
| Status | Torch Latency (ms) | Gems Latency (ms) | Speedup | TFLOPS | Shape |
|---|---|---|---|---|---|
| SUCCESS | 2.960 | 2.992 | **0.989x** | 0.718 | [1073741824] |
| SUCCESS | 0.005 | 0.006 | **0.960x** | 0.001 | [64, 64] |
| SUCCESS | 0.052 | 0.053 | **0.996x** | 0.638 | [4096, 4096] |
| SUCCESS | 0.053 | 0.053 | **0.996x** | 0.637 | [64, 512, 512] |
| SUCCESS | 3.028 | 2.996 | **1.010x** | 0.717 | [1024, 1024, 1024] |
| **Int32 Mean** | — | — | **~0.990x** | — | 5 cases, all PASS |

| **Overall Mean** | — | — | **~1.009x** | — | 10 cases |

## Multi-backend Testing
| Backend | Accuracy Test | Benchmark | Speedup (mean) | Notes |
|---|---|---|---|---|
| Nvidia (H20) | PASS | PASS (10 cases, --level core) | 1.009 | Primary |
| Tianshu | PASS | PASS (28 cases) | 0.996 | — |
| Muxi | PASS | PASS (28 cases) | 1.006 | — |
| Ascend | PASS | N/A | — | — |
| Hygon | PASS | PASS (28 cases) | 1.026 | — |

## Files Changed
- `src/flag_gems/ops/__irshift__.py`: Triton kernel implementation (pointwise_dynamic)
- `tests/test_irshift__.py`: Accuracy test (functional + inplace)
- `benchmark/test_irshift__.py`: Performance benchmark (BinaryPointwiseBenchmark)
- `src/flag_gems/ops/__init__.py`: Register import and `__all__`
- `src/flag_gems/__init__.py`: Register to `_FULL_CONFIG`
- `conf/operators.yaml`: Add operator entry (kind: Math, stage: alpha 5.1)

### CODE REVIEW

File: benchmark/test_irshift__.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import pytest
+import torch
+
+from . import base, consts
+
+
+@pytest.mark.irshift__
+def test___irshift___perf():
```
Comment (bwbwzzz) [inline benchmark/test_irshift__.py]: checking

File: tests/test_irshift__.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    ref_b = utils.to_reference(res_b)
+
+    # Use bitwise_right_shift as reference since __irshift__ has issues with broadcasting
+    ref_out = torch.bitwise_right_shift(ref_a, ref_b)
+    with flag_gems.use_gems():
+        res_out = torch.ops.aten.__irshift__(res_a, res_b)
+    utils.gems_assert_close(res_out, ref_out, dtype)
+
+
+@pytest.mark.irshift__
```
Comment (bwbwzzz) [inline tests/test_irshift__.py]: may need add inplace？

File: benchmark/test_irshift__.py (hunk 1 above)
Comment (bwbwzzz) [reply benchmark/test_irshift__.py]: 审查了 PR #3862，我觉得有两个需要改的实质问题：

__irshift__ 没有真正原地写回
在 src/flag_gems/ops/__irshift__.py:17-18：

def __irshift__(self, other, *, out=None):
    return __irshift__kernel(self, other, out=out)
aten::__irshift__ 的 schema 是 Tensor(a!) self -> Tensor(a!)，必须修改并返回同一个 self。项目里已有 inplace pointwise 写法是 out0=A，例如 bitwise_or_tensor_。这里应类似：

return __irshift__kernel(self, other, out0=self)
当前测试 res_a = res_a.__irshift__(res_b) 只是把变量重新绑定到返回值，没有检查原 tensor 是否被改写，所以会漏掉这个 bug。

缺少 Scalar 路径处理
PyTorch schema 有两个 overload：

aten::__irshift__.Tensor(Tensor(a!) self, Tensor other) -> Tensor(a!)
aten::__irshift__.Scalar(Tensor(a!) self, Scalar other) -> Tensor(a!)
但 kernel 默认把 a, b 都当 tensor。a >>= 1 / torch.ops.aten.__irshift__.Scalar(a, 1) 很可能走不到正确的 is_tensor=[True, False] wrapper。建议像 bitwise_and.py / bitwise_or.py 那样拆成 tensor 和 scalar 两个 kernel wrapper，并在注册里明确覆盖 Scalar/Tensor 路径，或至少补 scalar case 测试。

测试也建议补两点：检查 data_ptr/原始 res_a 内容确认真的 inplace；加入 scalar RHS，例如 res_a.__irshift__(3)。目前这版我不建议直接合。

**Review by bin913 — COMMENTED** [review]
## FlagGems Automated Review

**Operator:** `__irshift__`
**Verdict:** APPROVED (no format issues)

All structural checks passed — no naming, registration, test, or benchmark format issues found.

---
<sub>Automated review by FlagGems PR Review Skill</sub>

**Comment by bin913** [comment]
@bwbwzzz Please fix conflict

**Review by Dongxu-H — CHANGES_REQUESTED** [review]
## 🔍 Code Review

感谢您的贡献！这是一个简单的 in-place 右移操作符实现，但我发现了一些关键问题：

---

### 🚨 核心问题：In-place 语义不正确

**问题代码：**
```python
def __irshift__(self, other, *, out=None):
    logger.debug("GEMS __IRSHIFT__")
    return __irshift__kernel(self, other, out=out)
```

**PyTorch In-place 语义：**
```python
# a >>= b 应该直接修改 a，而不是创建新 tensor
a = torch.tensor([16, 32, 64])
b = torch.tensor([1, 2, 3])
a >>= b  # a 现在是 [8, 8, 8]，直接在原内存修改

# 等价于
torch.bitwise_right_shift(a, b, out=a)
# 或者
a.bitwise_right_shift_(b)
```

**当前实现的问题：**
- `pointwise_dynamic` kernel 默认创建新的输出 tensor
- 返回值是一个新 tensor，而不是修改 `self`
- 这破坏了 in-place 操作的核心语义

**正确的实现应该是：**
```python
def __irshift__(self, other):
    logger.debug("GEMS __IRSHIFT__")
    # In-place operation: result stored back in self
    return __irshift__kernel(self, other, out0=self)
```

注意：
- 使用 `out0=self` 而不是 `out=out`
- `__irshift__` 不应该接受 `out` 参数（这是 in-place 操作，输出永远是 self）

---

### ⚠️ 测试中的问题

**测试代码注释：**
```python
# Use bitwise_right_shift as reference since __irshift__ has issues with broadcasting
ref_out = torch.bitwise_right_shift(ref_a, ref_b)
```

**问题：**
1. 注释承认 "__irshift__ has issues with broadcasting"，但没有解决
2. 测试实际上没有验证 in-place 语义！它验证的是 `bitwise_right_shift`

**应该测试 in-place 行为：**
```python
@pytest.mark.irshift__
def test_irshift_inplace_semantics():
    # 验证 in-place 修改
    a = torch.randint(0, 100, (10,), dtype=torch.int32, device='cuda')
    a_id = id(a)
    a_data_ptr = a.data_ptr()
    
    b = torch.randint(0, 8, (10,), dtype=torch.int32, device='cuda')
    
    with flag_gems.use_gems():
        result = torch.ops.aten.__irshift__(a, b)
    
    # In-place 操作应该返回 self
    assert result is a  # 或者 assert id(result) == a_id
    # 内存地址应该不变
    assert result.data_ptr() == a_data_ptr
    # 结果应该正确
    ref = torch.bitwise_right_shift(a, b)
    assert torch.allclose(result, ref)
```

---

### 💡 参数验证缺失

```python
def __irshift__(self, other, *, out=None):
    # 缺少参数验证
```

**建议添加：**
```python
def __irshift__(self, other):
    logger.debug("GEMS __IRSHIFT__")
    
    # 参数验证
    if not torch.is_tensor(self):
        raise TypeError(f"Expected 'self' to be a tensor, but got {type(self)}")
    if not torch.is_tensor(other):
        # 支持标量
        other = torch.tensor(other, dtype=self.dtype, device=self.device)
    
    # 形状检查（in-place 不支持 broadcast 到更大形状）
    if self.shape != other.shape and other.shape != ():
        # PyTorch 会尝试广播，但结果必须 fit into self 的形状
        pass
    
    return __irshift__kernel(self, other, out0=self)
```

---

### 📊 测试覆盖建议

#### 1. 边界情况测试

```python
# 右移超过位数
x = torch.tensor([1], dtype=torch.int8, device='cuda')  # 8-bit integer
y = torch.tensor([10], dtype=torch.int8, device='cuda')  # shift by 10 bits
# 结果应该是 0（移位超过位宽）

# 负数右移（算术右移 vs 逻辑右移）
x = torch.tensor([-1, -2, -4], dtype=torch.int32, device='cuda')
y = torch.tensor([1, 1, 1], dtype=torch.int32, device='cuda')
# 算术右移：-1 >> 1 = -1, -2 >> 1 = -1, -4 >> 1 = -2
```

#### 2. In-place 验证测试

```python
def test_inplace_modification():
    """Verify that __irshift__ actually modifies self in-place."""
    a = torch.randint(0, 100, (100,), dtype=torch.int32, device='cuda')
    original_ptr = a.data_ptr()
    
    b = torch.randint(0, 8, (100,), dtype=torch.int32, device='cuda')
    
    with flag_gems.use_gems():
        result = torch.ops.aten.__irshift__(a, b)
    
    assert result.data_ptr() == original_ptr, "In-place operation should not allocate new memory"
    assert result is a, "In-place operation should return self"
```

---

### 🔧 实现建议

**推荐修正：**

```python
import logging
import torch
import triton
from flag_gems.utils import pointwise_dynamic

logger = logging.getLogger(__name__)

@pointwise_dynamic(promotion_methods=[(0, 1, "DEFAULT")])
@triton.jit
def irshift_kernel(a, b):
    return a >> b

def __irshift__(self, other):
    """
    In-place right shift: self >>= other
    
    Performs bitwise right shift in-place, modifying self.
    Equivalent to torch.bitwise_right_shift(self, other, out=self).
    """
    logger.debug("GEMS __IRSHIFT__")
    
    # Handle scalar other
    if not torch.is_tensor(other):
        other = torch.tensor(other, dtype=self.dtype, device=self.device)
    
    # In-place: store result back in self
    return irshift_kernel(self, other, out0=self)
```

---

### ✅ 做得好的地方

1. **代码简洁** - Kernel 实现简单明了
2. **注册正确** - operators.yaml, `__init__.py`, `_FULL_CONFIG` 都已更新
3. **PR 描述结构化** - 包含测试、性能数据、多后端结果
4. **CI 通过** - code-style, python-op 都通过

---

### 📋 后续步骤

**必须修复：**
1. ⚠️ 修正 in-place 语义 - 使用 `out0=self` 而不是返回新 tensor
2. 移除 `out` 参数 - `__irshift__` 不应该接受 out 参数
3. 添加 in-place 验证测试 - 确认返回的是 self，内存地址不变

**建议改进：**
4. 添加边界情况测试（大位移、负数）
5. 解决测试中注释提到的 broadcasting 问题
6. 解决合并冲突（@bin913 提到）

---

期待您的改进！🙏

**Comment by bwbwzzz** [comment]
> ### 📋 后续步骤
> **必须修复：**
> 
> 1. ⚠️ 修正 in-place 语义 - 使用 `out0=self` 而不是返回新 tensor
> 2. 移除 `out` 参数 - `__irshift__` 不应该接受 out 参数
> 3. 添加 in-place 验证测试 - 确认返回的是 self，内存地址不变
> 
> **建议改进：** 4. 添加边界情况测试（大位移、负数） 5. 解决测试中注释提到的 broadcasting 问题 6. 解决合并冲突（@bin913 提到）
> 
> 期待您的改进！🙏

have reviesed，I have updated the implementation accordingly. The in-place operation should now work as expected, and additional tests have been added.

**Review by Dongxu-H — DISMISSED** [review]
**[配置] ops/__init__.py 的 __all__ 未按字母顺序排列**
   
   \`__irshift__\` 放在了 \`SUPPORTED_FP8_DTYPE\` 之前，但按字母顺序应该放在 \`__ilshift__\` 之后：
   
   当前 PR 的顺序：
   \`\`\`python
   __all__ = [
       "__irshift__",         # 错误位置! 放在最前面
       "SUPPORTED_FP8_DTYPE",
       "ScaleDotProductAttention",
       "__ilshift__",
       ...
   ]
   \`\`\`
   
   正确顺序：
   \`\`\`python
   __all__ = [
       "SUPPORTED_FP8_DTYPE",  # S < _
       "ScaleDotProductAttention",
       "__ilshift__",          # __i-l < __i-r
       "__irshift__",          # 应放在 __ilshift__ 之后
       ...
   ]
   \`\`\`
   
   字母顺序：\`S\` (83) < \`_\` (95)，所以 \`SUPPORTED_*\` 应在 \`__*\` 之前
   
   另外：\`__ilshift__\` < \`__irshift__\` (l < r)，所以 \`__irshift__\` 应在 \`__ilshift__\` 之后

**Review by bin913 — DISMISSED** [review]

File: conf/operators.yaml [hunk 3]
```diff
@@ -3784,6 +3784,17 @@ ops:
       - stable: '2.2'
       - removed: '3.0'
       - beta: '5.3'
+  - id: irshift__
```
Comment (tengqm) [inline conf/operators.yaml]: I believe I have mentioned this... if there are double leading underscore AND double trailing underscore in operator name, the operator ID should drop both.

```suggestion
  - id: irshift
```

File: conf/operators.yaml [hunk 4]
```diff
@@ -3784,6 +3784,17 @@ ops:
...
+    description: Triton kernel implementation for __irshift__.
+    for:
+      - __irshift__
```
Comment (tengqm) [inline conf/operators.yaml]: For `ilshift`, the dispatch key was `__ilshift__.Tensor`, why there is not `.Tensor` here?

File: src/flag_gems/ops/__irshift__.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+@pointwise_dynamic(promotion_methods=[(0, 1, "DEFAULT")])
+@triton.jit
+def irshift_kernel(a, b):
+    return a >> b
+
+
+def __irshift__(self, other):
+    """In-place right shift: self >>= other."""
+    logger.debug("GEMS __IRSHIFT__")
```
Comment (tengqm) [inline src/flag_gems/ops/__irshift__.py]:
```suggestion
    logger.debug("GEMS IRSHIFT")
```

File: src/flag_gems/__init__.py [hunk 6]
```diff
@@ -40,6 +40,7 @@ def torch_ge(v):
     ("__ilshift__.Tensor", __ilshift__),
     ("__ior__.Scalar", bitwise_or_scalar_),
     ("__ior__.Tensor", bitwise_or_tensor_),
+    ("__irshift__", __irshift__),
```
Comment (tengqm) [inline src/flag_gems/__init__.py]: Do you feel strange when comparing this with line 40?

File: tests/test_irshift.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+# Shapes that support inplace bitwise operations with broadcasting
+INPLACE_BITWISE_SHAPES = [
+    ((512, 1024), (512, 1024)),
+    ((256, 512), (1, 512)),
+    ((256, 512), (256, 1)),
+    ((1024,), ()),
+]
+
+
+@pytest.mark.irshift__
```
Comment (tengqm) [inline tests/test_irshift.py]:
```suggestion
@pytest.mark.irshift
```

File: tests/test_irshift.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
+import pytest
```
Comment (tengqm) [inline tests/test_irshift.py:1]: Remove trailing underscores from the file name.

File: tests/test_irshift.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    ((256, 512), (1, 512)),
+    ((256, 512), (256, 1)),
+    ((1024,), ()),
+]
+
+
+@pytest.mark.irshift__
+@pytest.mark.parametrize("shapes", INPLACE_BITWISE_SHAPES)
+@pytest.mark.parametrize("dtype", utils.ALL_INT_DTYPES + [torch.uint8])
+def test_irshift__(shapes, dtype):
```
Comment (tengqm) [inline tests/test_irshift.py]:
```suggestion
def test_irshift(shapes, dtype):
```

File: tests/test_irshift.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+    with flag_gems.use_gems():
+        ret = res_a.__irshift__(res_b)
+
+    assert ret is res_a
+    assert res_a.data_ptr() == data_ptr
+    utils.gems_assert_close(res_a, ref_a, dtype)
+
+
+@pytest.mark.irshift__
```
Comment (tengqm) [inline tests/test_irshift.py]:
```suggestion
@pytest.mark.irshift
```

File: tests/test_irshift.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+@pytest.mark.parametrize("dtype", [torch.int8, torch.int16, torch.int32, torch.int64])
+def test_irshift__edge_cases(dtype):
```
Comment (tengqm) [inline tests/test_irshift.py]:
```suggestion
def test_irshift_edge_cases(dtype):
```

File: conf/operators.yaml (hunk 3 above)
Comment (bwbwzzz) [reply conf/operators.yaml]: I hadn't joined the PR work  yet when you previously mentioned this, but I have made the changes now. Thank you for the feedback!

File: tests/test_irshift.py (hunk 8 above)
Comment (bwbwzzz) [reply tests/test_irshift.py:1]: ok, have removed it, thanks

Comment (bwbwzzz) [reply tests/test_irshift.py:1]: i have reivised the file under benchmark

File: src/flag_gems/__init__.py (hunk 6 above)
Comment (bwbwzzz) [reply src/flag_gems/__init__.py]: yes , i have revised

**Review by Caeruleann — APPROVED** [review]
